File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/DatePickerDialogFragment.java
package com.remotearthsolutions.expensetracker.fragments;

import android.app.DatePickerDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.preference.PreferenceManager;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.utils.Constants;
import com.remotearthsolutions.expensetracker.utils.DateTimeUtils;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;


public class DatePickerDialogFragment extends DialogFragment implements View.OnClickListener {

    private LinearLayout previousdate, currentdate, selectdate;
    private TextView todayDateTv, yesterdayDateTv;
    private DatePickerDialogFragment.Callback callback;
    private int cDay, cMonth, cYear;
    private Context context;

    private String format;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.context = context;
    }

    public static DatePickerDialogFragment newInstance(String title) {
        DatePickerDialogFragment frag = new DatePickerDialogFragment();
        Bundle args = new Bundle();
        args.putString(Constants.KEY_TITLE, title);
        frag.setArguments(args);
        return frag;
    }

    public void setCallback(DatePickerDialogFragment.Callback callback) {
        this.callback = callback;
    }

    public void setInitialDate(int cDay, int cMonth, int cYear) {
        this.cDay = cDay;
        this.cMonth = cMonth;
        this.cYear = cYear;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.add_date, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getContext());
        format = preferences.getString(Constants.PREF_TIME_FORMAT,"dd-MM-yyyy");

        previousdate = view.findViewById(R.id.previousdate);
        currentdate = view.findViewById(R.id.currentdate);
        selectdate = view.findViewById(R.id.selectdate);
        yesterdayDateTv = view.findViewById(R.id.showdyesterday);
        todayDateTv = view.findViewById(R.id.showdtoday);

        yesterdayDateTv.setText(DateTimeUtils.getDate(format, -1));
        todayDateTv.setText(DateTimeUtils.getCurrentDate(format));

        previousdate.setOnClickListener(this);
        currentdate.setOnClickListener(this);
        selectdate.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.currentdate: {
                String date = DateTimeUtils.getCurrentDate(format);
                callback.onSelectDate(date);
                break;
            }
            case R.id.previousdate: {
                String date = DateTimeUtils.getDate(format, -1);
                callback.onSelectDate(date);
                break;
            }
            case R.id.selectdate: {
                DatePickerDialog datePickerDialog = new DatePickerDialog(context, (view, year, month, dayOfMonth) -> {
                    cYear = year;
                    cMonth = month;
                    cDay = dayOfMonth;
                    Calendar calendar = Calendar.getInstance();
                    calendar.set(Calendar.DAY_OF_MONTH, cDay);
                    calendar.set(Calendar.MONTH, cMonth);
                    calendar.set(Calendar.YEAR, cYear);
                    DateFormat dateFormat = new SimpleDateFormat(format, Locale.getDefault());

                    callback.onSelectDate(dateFormat.format(calendar.getTime()));
                    dismiss();
                }, cYear, cMonth, cDay);
                datePickerDialog.show();
                break;
            }
        }
    }

    public interface Callback {
        void onSelectDate(String date);
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/fragments/DatePickerDialogFragment.kt
package com.remotearthsolutions.expensetracker.fragments

import android.app.DatePickerDialog
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.DatePicker
import androidx.fragment.app.DialogFragment
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.databinding.AddDateBinding
import com.remotearthsolutions.expensetracker.utils.Constants
import com.remotearthsolutions.expensetracker.utils.DateTimeUtils.getCurrentDate
import com.remotearthsolutions.expensetracker.utils.DateTimeUtils.getDate
import com.remotearthsolutions.expensetracker.utils.SharedPreferenceUtils
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class DatePickerDialogFragment : DialogFragment(),
    View.OnClickListener {
    private lateinit var binding: AddDateBinding
    private var callback: Callback? = null
    private var cDay = 0
    private var cMonth = 0
    private var cYear = 0
    private lateinit var mContext: Context

    private lateinit var format: String
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    fun setInitialDate(cDay: Int, cMonth: Int, cYear: Int) {
        this.cDay = cDay
        this.cMonth = cMonth
        this.cYear = cYear
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = AddDateBinding.inflate(layoutInflater, container, false)
        return binding.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        format = SharedPreferenceUtils.getInstance(requireActivity())?.getString(
            Constants.PREF_TIME_FORMAT,
            Constants.KEY_DATE_MONTH_YEAR_DEFAULT
        )!!
        binding.showdyesterday.text = getDate(format, -1)
        binding.showdtoday.text = getCurrentDate(format)
        binding.previousdate.setOnClickListener(this)
        binding.currentdate.setOnClickListener(this)
        binding.selectdate.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.currentdate -> {
                val date = getCurrentDate(format)
                callback!!.onSelectDate(date)
            }
            R.id.previousdate -> {
                val date = getDate(format, -1)
                callback!!.onSelectDate(date)
            }
            R.id.selectdate -> {
                val datePickerDialog = DatePickerDialog(
                    mContext,
                    { _: DatePicker?, year: Int, month: Int, dayOfMonth: Int ->
                        cYear = year
                        cMonth = month
                        cDay = dayOfMonth
                        val calendar = Calendar.getInstance()
                        calendar[Calendar.DAY_OF_MONTH] = cDay
                        calendar[Calendar.MONTH] = cMonth
                        calendar[Calendar.YEAR] = cYear
                        val dateFormat: DateFormat = SimpleDateFormat(
                            format,
                            Locale.getDefault()
                        )
                        callback!!.onSelectDate(dateFormat.format(calendar.time))
                        dismiss()
                    },
                    cYear,
                    cMonth,
                    cDay
                )
                datePickerDialog.show()
            }
        }
    }

    interface Callback {
        fun onSelectDate(date: String?)
    }

    companion object {
        fun newInstance(title: String?): DatePickerDialogFragment {
            val frag = DatePickerDialogFragment()
            val args = Bundle()
            args.putString(Constants.KEY_TITLE, title)
            frag.arguments = args
            return frag
        }
    }
}